File: internal/browser/snapshot.go
package browser

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/sirhap/piercode/internal/tool"
)

const (
	defaultMaxSnapshotNodes = 200
	defaultMaxSnapshotDepth = 15
	maxSnapshotOutputChars  = 12000
	maxSnapshotValueChars   = 120
)

type axTreeResponse struct {
	Nodes []axNode `json:"nodes"`
}

type axNode struct {
	NodeID       string       `json:"nodeId"`
	BackendDOMID int          `json:"backendDOMNodeId"`
	ChildIDs     []string     `json:"childIds"`
	ParentID     string       `json:"parentId"`
	Role         axValue      `json:"role"`
	Name         axValue      `json:"name"`
	Value        axValue      `json:"value"`
	Description  axValue      `json:"description"`
	Ignored      bool         `json:"ignored"`
	Properties   []axProperty `json:"properties"`
}

type axProperty struct {
	Name  string  `json:"name"`
	Value axValue `json:"value"`
}

type axValue struct {
	Type  string          `json:"type"`
	Value json.RawMessage `json:"value"`
}

// CompactSnapshot renders the accessibility tree as an indented hierarchy.
// Parent-child context is preserved (each kept node is indented under its
// nearest kept ancestor), refs (e0, e1, …) are assigned to actionable nodes,
// and output is bounded by opts.MaxNodes / opts.MaxChars / opts.Depth. When
// opts.RefID is set, only that ref's subtree is rendered. The legacy flat
// behavior is gone; refs remain document-ordered so they stay stable.
func CompactSnapshot(raw json.RawMessage, tab tool.BrowserTab, snapshotID string, opts tool.SnapshotOptions) (tool.BrowserSnapshot, []RefTarget, error) {
	maxNodes := opts.MaxNodes
	if maxNodes <= 0 {
		maxNodes = defaultMaxSnapshotNodes
	}
	maxChars := opts.MaxChars
	if maxChars <= 0 {
		maxChars = maxSnapshotOutputChars
	}
	maxDepth := opts.Depth
	if maxDepth <= 0 {
		maxDepth = defaultMaxSnapshotDepth
	}

	var tree axTreeResponse
	if err := json.Unmarshal(raw, &tree); err != nil {
		return tool.BrowserSnapshot{}, nil, err
	}

	// Index nodes by id and find roots (nodes with no parent in the set).
	byID := make(map[string]*axNode, len(tree.Nodes))
	for i := range tree.Nodes {
		byID[tree.Nodes[i].NodeID] = &tree.Nodes[i]
	}
	var roots []*axNode
	for i := range tree.Nodes {
		n := &tree.Nodes[i]
		if n.ParentID == "" || byID[n.ParentID] == nil {
			roots = append(roots, n)
		}
	}
	// If RefID is requested, re-root at the node whose ref maps to it. We need a
	// first pass to assign refs in document order so RefID resolves to the same
	// node a prior snapshot returned.

	var out strings.Builder
	fmt.Fprintf(&out, "snapshotId=%s url=%q title=%q\n\n", snapshotID, tab.URL, tab.SafeTitle())
	refs := make([]RefTarget, 0)
	state := &snapshotWalk{
		out:      &out,
		refs:     &refs,
		maxNodes: maxNodes,
		maxChars: maxChars,
		maxDepth: maxDepth,
		byID:     byID,
	}

	// Determine the walk roots. When RefID is set we still need to assign refs
	// consistently, so we walk the whole tree but only EMIT the matching
	// subtree. Simplicity beats cleverness: assign refs during a structural
	// walk, and gate emission on whether we're inside the requested subtree.
	state.targetRef = strings.TrimSpace(opts.RefID)
	state.emitting = state.targetRef == "" // emit from the top unless filtering

	for _, root := range roots {
		state.walk(root, 0)
	}

	if state.targetRef != "" && !state.matchedRef {
		return tool.BrowserSnapshot{}, nil, fmt.Errorf("ref %q not found in this snapshot; take a fresh browser_snapshot", state.targetRef)
	}

	if state.truncated {
		fmt.Fprintf(&out, "\n... snapshot truncated at %d nodes / %d chars. Narrow it: pass a smaller depth, or focus a subtree with ref_id=<ref>.\n", maxNodes, maxChars)
	}
	return tool.BrowserSnapshot{
		SnapshotID: snapshotID,
		Tab:        tab,
		Text:       strings.TrimSpace(out.String()),
		NodeCount:  len(tree.Nodes),
		RefCount:   len(refs),
		Truncated:  state.truncated,
	}, refs, nil
}

type snapshotWalk struct {
	out        *strings.Builder
	refs       *[]RefTarget
	byID       map[string]*axNode
	maxNodes   int
	maxChars   int
	maxDepth   int
	shown      int
	truncated  bool
	targetRef  string // when non-empty, only emit this ref's subtree
	emitting   bool   // currently inside the emit scope
	matchedRef bool   // the targetRef was found
}

// walk descends the AX subtree rooted at node. visualDepth is the indentation
// level among KEPT nodes (so skipped wrapper nodes don't inflate indentation).
func (s *snapshotWalk) walk(node *axNode, visualDepth int) {
	if s.truncated || node == nil {
		return
	}
	role := node.Role.String()
	name := trimText(node.Name.String(), maxSnapshotValueChars)
	value := trimText(node.Value.String(), maxSnapshotValueChars)
	desc := trimText(node.Description.String(), maxSnapshotValueChars)
	keep := shouldKeepAXNode(*node, role, name, value, desc)

	childDepth := visualDepth
	if keep {
		// Assign a ref to actionable nodes regardless of emit scope, so refs
		// stay stable whether or not a subtree filter is active.
		refName := ""
		if isImportantRole(role) || isFocusable(*node) || isEditable(*node) {
			refName = fmt.Sprintf("e%d", len(*s.refs))
			*s.refs = append(*s.refs, RefTarget{
				Ref:       refName,
				NodeID:    node.NodeID,
				BackendID: node.BackendDOMID,
				Role:      role,
				Name:      name,
				Bounds:    boundsFromNode(*node),
			})
		}

		// Subtree filtering: enter emit scope when we hit the target ref.
		enteredHere := false
		if s.targetRef != "" && !s.emitting && refName == s.targetRef {
			s.emitting = true
			s.matchedRef = true
			enteredHere = true
		}

		if s.emitting {
			if s.shown >= s.maxNodes || s.out.Len() >= s.maxChars {
				s.truncated = true
				return
			}
			s.writeNode(node, role, name, value, desc, refName, visualDepth)
			s.shown++
			childDepth = visualDepth + 1
		}

		// Stop descending past the depth cap (within emit scope).
		if s.emitting && childDepth > s.maxDepth {
			if enteredHere {
				s.emitting = false
			}
			return
		}

		for _, cid := range node.ChildIDs {
			s.walk(s.byID[cid], childDepth)
			if s.truncated {
				break
			}
		}
		if enteredHere {
			s.emitting = false // leave the requested subtree
		}
		return
	}

	// Node not kept: descend without indenting or emitting it.
	for _, cid := range node.ChildIDs {
		s.walk(s.byID[cid], visualDepth)
		if s.truncated {
			break
		}
	}
}

func (s *snapshotWalk) writeNode(node *axNode, role, name, value, desc, refName string, depth int) {
	out := s.out
	for i := 0; i < depth; i++ {
		out.WriteString("  ")
	}
	if refName != "" {
		out.WriteString("[")
		out.WriteString(refName)
		out.WriteString("] ")
	} else {
		out.WriteString("- ")
	}
	out.WriteString(role)
	if name != "" {
		fmt.Fprintf(out, " %q", name)
	}
	if value != "" && value != name {
		fmt.Fprintf(out, " value=%q", value)
	}
	if desc != "" && desc != name {
		fmt.Fprintf(out, " desc=%q", desc)
	}
	for _, flag := range compactFlags(*node) {
		out.WriteByte(' ')
		out.WriteString(flag)
	}
	out.WriteByte('\n')
}

func (v axValue) String() string {
	if len(v.Value) == 0 || string(v.Value) == "null" {
		return ""
	}
	var s string
	if err := json.Unmarshal(v.Value, &s); err == nil {
		return s
	}
	var b bool
	if err := json.Unmarshal(v.Value, &b); err == nil {
		if b {
			return "true"
		}
		return "false"
	}
	var n float64
	if err := json.Unmarshal(v.Value, &n); err == nil {
		return fmt.Sprintf("%g", n)
	}
	return strings.Trim(string(v.Value), `"`)
}

func shouldKeepAXNode(node axNode, role, name, value, desc string) bool {
	if node.Ignored {
		return false
	}
	if role == "" || role == "none" || role == "generic" {
		return false
	}
	if name == "" && value == "" && desc == "" && !isImportantRole(role) && !isFocusable(node) && !isEditable(node) {
		return false
	}
	return true
}

func isImportantRole(role string) bool {
	switch strings.ToLower(role) {
	case "button", "textbox", "searchbox", "link", "checkbox", "radio", "combobox",
		"menuitem", "tab", "heading", "slider", "spinbutton", "switch", "listbox",
		"option", "treeitem", "gridcell", "cell":
		return true
	default:
		return false
	}
}

func compactFlags(node axNode) []string {
	flags := make([]string, 0, 6)
	for _, prop := range node.Properties {
		name := strings.ToLower(prop.Name)
		value := prop.Value.String()
		switch name {
		case "disabled", "focused", "selected", "checked", "pressed", "expanded":
			if value != "" && value != "false" {
				flags = append(flags, name+"="+value)
			}
		case "editable":
			if value != "" && value != "false" {
				flags = append(flags, "editable")
			}
		case "level":
			if value != "" {
				flags = append(flags, "level="+value)
			}
		case "url":
			if value != "" {
				flags = append(flags, "href="+fmt.Sprintf("%q", trimText(value, maxSnapshotValueChars)))
			}
		}
	}
	return flags
}

func isFocusable(node axNode) bool {
	for _, prop := range node.Properties {
		if strings.EqualFold(prop.Name, "focusable") && prop.Value.String() == "true" {
			return true
		}
	}
	return false
}

func isEditable(node axNode) bool {
	for _, prop := range node.Properties {
		if strings.EqualFold(prop.Name, "editable") {
			v := prop.Value.String()
			return v != "" && v != "false"
		}
	}
	return false
}

func boundsFromNode(node axNode) *Bounds {
	for _, prop := range node.Properties {
		if !strings.EqualFold(prop.Name, "bounds") || len(prop.Value.Value) == 0 {
			continue
		}
		var b Bounds
		if err := json.Unmarshal(prop.Value.Value, &b); err == nil && b.Width > 0 && b.Height > 0 {
			return &b
		}
	}
	return nil
}

func trimText(s string, limit int) string {
	s = strings.Join(strings.Fields(strings.TrimSpace(s)), " ")
	if limit <= 0 || len([]rune(s)) <= limit {
		return s
	}
	runes := []rune(s)
	return string(runes[:limit]) + "..."
}

File: extension/src/background/browser/snapshot.ts
// Faithful port of internal/browser/snapshot.go — AX tree → indented text + refs.
// Parent-child context preserved (indent under nearest kept ancestor), refs (e0,e1…)
// assigned to actionable nodes in document order, bounded by maxNodes/maxChars/depth.
//
// ── #7 (part 2): serialized-snapshot token footprint vs a compact format ────────
// One emitted node line ≈ `  [e3] button "Submit Order" focused\n`. Cost per line:
//   • indent: 2 spaces × depth (≈0.5 tok at depth 5–8 once GPT-BPE merges runs),
//   • ref tag `[eN] ` ≈ 3 tok, role word ≈ 1 tok, quoted name ≈ name_words+2 tok,
//   • each affordance flag ≈ 1–2 tok.
// A typical 80–150-node page renders ~120–500 lines ⇒ ~600–2500 tokens, hard-capped
// by MAX_OUTPUT_CHARS=12000 (~3000 tok) + the 200-node cap. That is ALREADY the AX-text
// "DOM mode" the openchrome reference touts (5–15× vs raw HTML): one terse line/node,
// no markup, no inline styles. The cheap further wins would be 1-space indent and
// dropping the `value=`/`desc=` key prefixes — but those DIVERGE from the Go port
// (internal/browser/snapshot.go) this file mirrors byte-for-byte and would desync the
// two serializers + their tests, so they are deliberately NOT applied here. The two
// real caps (maxChars + maxNodes, with the `ref_id=<ref>`/`depth` narrowing hint on
// truncation) are the load-bearing token guards and are already in place. #7 part-2 is
// therefore a documented finding, not a code change. (Part 1 — stable backendNodeId
// addressing — is verified in ref-resolve.ts and IS satisfied.)
import type { BrowserTab, RefTarget, Bounds } from './types'

const DEFAULT_MAX_NODES = 200
const DEFAULT_MAX_DEPTH = 15
const MAX_OUTPUT_CHARS = 12000
const MAX_VALUE_CHARS = 120

export interface SnapshotOptions {
  refId?: string
  maxNodes?: number
  maxChars?: number
  depth?: number
}

export interface FrameAXTree { raw: AXTreeResponse; sessionId: string; url: string }

interface AXValueRaw { type?: string; value?: any }
interface AXProperty { name: string; value?: AXValueRaw }
interface AXNode {
  nodeId: string
  backendDOMNodeId?: number
  childIds?: string[]
  parentId?: string
  role?: AXValueRaw
  name?: AXValueRaw
  value?: AXValueRaw
  description?: AXValueRaw
  ignored?: boolean
  properties?: AXProperty[]
}
export interface AXTreeResponse { nodes: AXNode[] }

export interface SnapshotResult {
  text: string
  refs: Record<string, RefTarget>
  nodeCount: number
  refCount: number
  truncated: boolean
}

// --- axValue.String() port ---
function axStr(v?: AXValueRaw): string {
  if (!v || v.value == null) return ''
  const x = v.value
  if (typeof x === 'string') return x
  if (typeof x === 'boolean') return x ? 'true' : 'false'
  if (typeof x === 'number') return `${x}`
  return String(x).replace(/^"|"$/g, '')
}

function trimText(s: string, limit: number): string {
  s = s.trim().split(/\s+/).filter(Boolean).join(' ')
  const runes = [...s]
  if (limit <= 0 || runes.length <= limit) return s
  return runes.slice(0, limit).join('') + '...'
}

function quote(s: string): string { return JSON.stringify(s) }

const IMPORTANT_ROLES = new Set([
  'button', 'textbox', 'searchbox', 'link', 'checkbox', 'radio', 'combobox',
  'menuitem', 'tab', 'heading', 'slider', 'spinbutton', 'switch', 'listbox',
  'option', 'treeitem', 'gridcell', 'cell',
])
function isImportantRole(role: string): boolean { return IMPORTANT_ROLES.has(role.toLowerCase()) }

function isFocusable(n: AXNode): boolean {
  return (n.properties ?? []).some(p => p.name.toLowerCase() === 'focusable' && axStr(p.value) === 'true')
}
function isEditable(n: AXNode): boolean {
  for (const p of n.properties ?? []) {
    if (p.name.toLowerCase() === 'editable') { const v = axStr(p.value); return v !== '' && v !== 'false' }
  }
  return false
}
function shouldKeep(n: AXNode, role: string, name: string, value: string, desc: string): boolean {
  if (n.ignored) return false
  if (role === '' || role === 'none' || role === 'generic') return false
  if (name === '' && value === '' && desc === '' && !isImportantRole(role) && !isFocusable(n) && !isEditable(n)) return false
  return true
}
function boundsFromNode(n: AXNode): Bounds | null {
  for (const p of n.properties ?? []) {
    if (p.name.toLowerCase() !== 'bounds' || p.value?.value == null) continue
    const b = p.value.value as Bounds
    if (b && b.width > 0 && b.height > 0) return b
  }
  return null
}
function compactFlags(n: AXNode): string[] {
  const flags: string[] = []
  for (const p of n.properties ?? []) {
    const name = p.name.toLowerCase()
    const value = axStr(p.value)
    switch (name) {
      case 'disabled': case 'focused': case 'selected': case 'checked': case 'pressed': case 'expanded':
        if (value !== '' && value !== 'false') flags.push(`${name}=${value}`); break
      case 'editable':
        if (value !== '' && value !== 'false') flags.push('editable'); break
      case 'level':
        if (value !== '') flags.push(`level=${value}`); break
      case 'url':
        if (value !== '') flags.push(`href=${quote(trimText(value, MAX_VALUE_CHARS))}`); break
    }
  }
  return flags
}

class Walk {
  out = ''
  refs: RefTarget[] = []
  shown = 0
  truncated = false
  emitting: boolean
  matchedRef = false
  constructor(
    private byId: Map<string, AXNode>,
    private maxNodes: number, private maxChars: number, private maxDepth: number,
    private targetRef: string, private refBase: number,
  ) { this.emitting = targetRef === '' }

  walk(node: AXNode | undefined, visualDepth: number): void {
    if (this.truncated || !node) return
    const role = axStr(node.role)
    const name = trimText(axStr(node.name), MAX_VALUE_CHARS)
    const value = trimText(axStr(node.value), MAX_VALUE_CHARS)
    const desc = trimText(axStr(node.description), MAX_VALUE_CHARS)
    const keep = shouldKeep(node, role, name, value, desc)
    let childDepth = visualDepth

    if (keep) {
      let refName = ''
      if (isImportantRole(role) || isFocusable(node) || isEditable(node)) {
        // #7 (part 1): the ref NAME (e0,e1…) is positional within THIS snapshot and is
        // re-minted on every re-snapshot, but the stored RefTarget carries the CDP
        // `backendDOMNodeId` — a stable per-document node handle. resolvePoint() resolves
        // a ref THROUGH that backendId (boxModelBounds), so the same element addresses
        // consistently across calls after a refresh; the name is just a lookup label, the
        // backendId is the real address. (nodeId is the volatile AX id — not used to act.)
        refName = `e${this.refBase + this.refs.length}`
        this.refs.push({ ref: refName, nodeId: node.nodeId, backendId: node.backendDOMNodeId ?? 0,
          role, name, bounds: boundsFromNode(node), sessionId: '', frameOffset: null })
      }
      let enteredHere = false
      if (this.targetRef !== '' && !this.emitting && refName === this.targetRef) {
        this.emitting = true; this.matchedRef = true; enteredHere = true
      }
      if (this.emitting) {
        if (this.shown >= this.maxNodes || this.out.length >= this.maxChars) { this.truncated = true; return }
        this.writeNode(node, role, name, value, desc, refName, visualDepth)
        this.shown++
        childDepth = visualDepth + 1
      }
      if (this.emitting && childDepth > this.maxDepth) { if (enteredHere) this.emitting = false; return }
      for (const cid of node.childIds ?? []) { this.walk(this.byId.get(cid), childDepth); if (this.truncated) break }
      if (enteredHere) this.emitting = false
      return
    }
    for (const cid of node.childIds ?? []) { this.walk(this.byId.get(cid), visualDepth); if (this.truncated) break }
  }

  private writeNode(node: AXNode, role: string, name: string, value: string, desc: string, refName: string, depth: number): void {
    let line = '  '.repeat(depth)
    line += refName !== '' ? `[${refName}] ` : '- '
    line += role
    if (name !== '') line += ` ${quote(name)}`
    if (value !== '' && value !== name) line += ` value=${quote(value)}`
    if (desc !== '' && desc !== name) line += ` desc=${quote(desc)}`
    for (const flag of compactFlags(node)) line += ` ${flag}`
    this.out += line + '\n'
  }
}

function indexAndRoots(tree: AXTreeResponse): { byId: Map<string, AXNode>; roots: AXNode[] } {
  const byId = new Map<string, AXNode>()
  for (const n of tree.nodes) byId.set(n.nodeId, n)
  const roots: AXNode[] = []
  for (const n of tree.nodes) if (!n.parentId || !byId.has(n.parentId)) roots.push(n)
  return { byId, roots }
}

function refsToMap(arr: RefTarget[]): Record<string, RefTarget> {
  const m: Record<string, RefTarget> = {}
  for (const r of arr) m[r.ref] = r
  return m
}

export function compactSnapshot(
  raw: AXTreeResponse, tab: BrowserTab, snapshotId: string, opts: SnapshotOptions = {},
): SnapshotResult {
  const maxNodes = opts.maxNodes && opts.maxNodes > 0 ? opts.maxNodes : DEFAULT_MAX_NODES
  const maxChars = opts.maxChars && opts.maxChars > 0 ? opts.maxChars : MAX_OUTPUT_CHARS
  const maxDepth = opts.depth && opts.depth > 0 ? opts.depth : DEFAULT_MAX_DEPTH
  const { byId, roots } = indexAndRoots(raw)
  const targetRef = (opts.refId ?? '').trim()

  const w = new Walk(byId, maxNodes, maxChars, maxDepth, targetRef, 0)
  let header = `snapshotId=${snapshotId} url=${quote(tab.url)} title=${quote(tab.title)}\n\n`
  for (const r of roots) w.walk(r, 0)
  if (targetRef !== '' && !w.matchedRef) {
    throw new Error(`ref ${quote(targetRef)} not found in this snapshot; take a fresh browser_snapshot`)
  }
  let text = header + w.out
  if (w.truncated) {
    text += `\n... snapshot truncated at ${maxNodes} nodes / ${maxChars} chars. Narrow it: pass a smaller depth, or focus a subtree with ref_id=<ref>.\n`
  }
  return { text: text.trim(), refs: refsToMap(w.refs), nodeCount: raw.nodes.length, refCount: w.refs.length, truncated: w.truncated }
}

// renderFrameTree: one OOPIF frame, refs starting at refBase, indented under header.
function renderFrameTree(raw: AXTreeResponse, refBase: number, opts: SnapshotOptions): { text: string; refs: RefTarget[] } {
  const maxNodes = opts.maxNodes && opts.maxNodes > 0 ? opts.maxNodes : DEFAULT_MAX_NODES
  const maxChars = opts.maxChars && opts.maxChars > 0 ? opts.maxChars : MAX_OUTPUT_CHARS
  const maxDepth = opts.depth && opts.depth > 0 ? opts.depth : DEFAULT_MAX_DEPTH
  const { byId, roots } = indexAndRoots(raw)
  const w = new Walk(byId, maxNodes, maxChars, maxDepth, '', refBase)
  for (const r of roots) w.walk(r, 1)
  return { text: w.out, refs: w.refs }
}

export function compactSnapshotWithFrames(
  raw: AXTreeResponse, frames: FrameAXTree[], tab: BrowserTab, snapshotId: string, opts: SnapshotOptions = {},
): SnapshotResult {
  const base = compactSnapshot(raw, tab, snapshotId, opts)
  if ((opts.refId ?? '') !== '' || frames.length === 0) return base
  let out = base.text
  const refs: RefTarget[] = Object.values(base.refs)
  for (const fr of frames) {
    const { text: ftext, refs: frefs } = renderFrameTree(fr.raw, refs.length, { ...opts, refId: '' })
    if (frefs.length === 0) continue
    out += `\n\n[iframe (cross-origin) ${trimText(fr.url, 120)}]\n` + ftext
    for (const r of frefs) r.sessionId = fr.sessionId
    refs.push(...frefs)
  }
  return { ...base, text: out.trim(), refs: refsToMap(refs), refCount: refs.length }
}
